Tópico 2 - Problema da Mochila (Knapsack Problem)

O Problema da Mochila (Knapsack Problem) é um problema clássico de otimização em ciência da computação e matemática. Trata-se de um problema de otimização combinatória em que o objetivo é selecionar um subconjunto de itens com pesos e valores dados, sujeitos a uma restrição no peso total. O objetivo é maximizar o valor total dos itens selecionados.
O problema é frequentemente descrito no contexto de uma mochila com capacidade fixa. Cada item tem um peso específico e um valor correspondente. O desafio é determinar a combinação de itens a incluir na mochila, garantindo que o peso total não ultrapasse a capacidade, ao mesmo tempo que se maximiza o valor total.
O problema pode ser formalmente definido da seguinte forma:
•	Dado um conjunto de itens, cada um com um peso wi e um valor vi, onde i representa o índice do item.
•	Dada uma mochila com capacidade W.
•	Encontrar um subconjunto de itens que maximize o valor total ∑ivi, mantendo o peso total ∑iwi abaixo ou igual à capacidade da mochila.
Existem diferentes variações do Problema da Mochila, incluindo o Problema da Mochila 0/1 (onde cada item pode ser incluído ou excluído). É esta variação que iremos usar neste trabalho.
O Problema da Mochila é conhecido por ser NP-difícil, o que significa que não existe um algoritmo conhecido de tempo polinomial para resolvê-lo exatamente para todas as instâncias. No entanto, diversos algoritmos, heurísticas e métodos de aproximação são utilizados para encontrar soluções aproximadas de forma eficiente.


Para resolver o problema descrito acima, iremos recorrer ao algoritmo Simulated Annealing.

1. Codifica um algoritmo capaz de fazer uma procura pelo ótimo global:

In [ ]:
#################   1.Codifica um algoritmo capaz de fazer uma procura pelo ótimo global   ###################

import numpy as np

# Criação do dataset
def create_ksp(number_of_items, max_utility=100, max_weight=100):
    utility = np.random.randint(0, max_utility, number_of_items)
    capacity = np.random.randint(1, max_weight, number_of_items)
    return utility, capacity

print("Valor dos itens:", utility)
print("Peso dos itens:", capacity)

Valor dos itens [ 9 15 64 28 89 93 29  8 73  0 40 36 16 11 54 88 62 33 72 78 49 51 54 77
 69 13 25 13 92 86 30 30 89 12 65 31 57 36 27 18 93 77 22 23 94 11 28 74
 88  9 15 18 80 71 88 11 17 46  7 75 28 33 84 96 88 44  5  4 71 88 88 50
 54 34 15 77 88 15  6 85 22 11 12 92 96 62 57 79 42 57 97 50 45 40 89 73
 37  0 18 23]
Peso dos itens [25 88 59 40 28 14 44 64 88 70  8 87  0  7 87 62 10 80  7 34 34 32  4 40
 27  6 72 71 11 33 32 47 22 61 87 36 98 43 85 90 34 64 98 46 77  2  0  4
 89 13 26  8 78 14 89 41 76 50 62 95 51 95  3 93 22 14 42 28 35 12 31 70
 58 85 27 65 41 44 61 56  5 27 27 43 83 29 61 74 91 88 61 96  0 26 61 76
  2 69 71 26]


In [ ]:
#################   1.Codifica um algoritmo capaz de fazer uma procura pelo ótimo global (continuação)  ###################


# Define a Função objetivo
def objective_function(utility, capacity, max_capacity, choice):
    total_weight = np.dot(capacity, choice)
    if total_weight <= max_capacity:
        return np.dot(utility, choice)
    else:
        return 0

# Função de perturbação para o Simulated Annealing
def perturb_solution(solution):
    perturbed_solution = solution.copy()
    indices_to_change = np.random.choice(len(solution), size=int(0.1 * len(solution)), replace=False)
    perturbed_solution[indices_to_change] = 1 - perturbed_solution[indices_to_change]
    return perturbed_solution

# Função para o Simulated Annealing
def simulated_annealing(initial_solution, cost_function, temperature, cooling_rate, iterations):
    current_solution = initial_solution
    current_cost = cost_function(current_solution)

    best_solution = current_solution
    best_cost = current_cost

    temperature_list = []

    for i in range(iterations):
        temperature_list.append(temperature)

        #Gera uma solução perturbada
        next_solution = perturb_solution(current_solution)
        next_cost = cost_function(next_solution)

        #Aceita a solução perturbada com uma dada probabilidade
        if next_cost > current_cost or np.random.rand() < np.exp((current_cost - next_cost) / temperature):
            current_solution = next_solution
            current_cost = next_cost

            if current_cost > best_cost:
                best_solution = current_solution
                best_cost = current_cost

        temperature *= cooling_rate

    return best_solution, best_cost, temperature_list

# Parâmetros do Simulated Annealing
np.random.seed(42)  # Define a seed do gerador de números aleatórios
initial_temperature = 1000.0
cooling_rate = 0.99
iterations = 1000

# Executando o Simulated Annealing
utility, capacity = create_ksp(100)
choice = np.random.choice([0, 1], 100)
max_capacity = 2500

# Função de custo sem negação
def cost_function(solution):
    return objective_function(utility, capacity, max_capacity, solution)

best_solution, best_cost, temperature_list = simulated_annealing(
    choice, cost_function,
    initial_temperature, cooling_rate, iterations
)

print("Melhor solução encontrada:", best_solution)
print("Valor da melhor solução:", best_cost)
total_weight_of_best_solution = np.dot(capacity, best_solution)
print("Peso total da melhor solução:", total_weight_of_best_solution)

Melhor solução encontrada: [1 1 0 0 0 0 0 0 1 0 1 1 0 1 0 1 1 0 0 1 0 1 1 0 1 0 0 0 0 1 0 1 1 1 1 1 1
 1 1 0 1 1 0 1 1 1 0 0 1 1 0 0 1 0 0 0 0 0 0 1 1 1 1 1 0 1 1 1 1 1 0 0 0 0
 1 0 0 1 1 1 1 0 1 0 0 0 0 1 0 0 1 1 1 1 0 1 0 1 0 0]
Valor da melhor solução: 3103
Peso total da melhor solução: 2477


C:\Users\Carlos\AppData\Local\Temp\ipykernel_24556\4179523337.py:34: RuntimeWarning: overflow encountered in exp
  if next_cost > current_cost or np.random.rand() < np.exp((current_cost - next_cost) / temperature):


A melhor solução encontrada é uma escolha binária de itens representada por um array de valores de 0 e 1.
O valor total (3103) e o peso total (2477) representam os valores para a melhor solução, não excedendo o peso máximo de 2500.
Estes resultados mostram a aplicação bem-sucedida do algoritmo Simulated Annealing para o Problema da Mochila, encontrando uma solução que maximiza o valor total dos itens, respeitando a capacidade da mochila.

2. Testa várias combinações de Temperatura Inicial e de método de atualização de Temperatura, e comenta sobre o impacto que esta atualização e valor tem no resultado final do modelo. 

In [ ]:
#################   2. Testa várias combinações de Temperatura Inicial e de método de atualização de Temperatura, #################
#################   e comenta sobre o impacto que esta atualização e valor tem no resultado final do modelo.    ###################
    
import numpy as np

# Criação do dataset
def create_ksp(number_of_items, max_utility=100, max_weight=100):
    utility = np.random.randint(0, max_utility, number_of_items)
    capacity = np.random.randint(0, max_weight, number_of_items)
    return utility, capacity

# Função objetivo
def objective_function(utility, capacity, max_capacity, choice):
    total_weight = np.dot(capacity, choice)
    if total_weight <= max_capacity:
        return np.dot(utility, choice)
    else:
        return 0

# Função de perturbação para o Simulated Annealing
def perturb_solution(solution):
    perturbed_solution = solution.copy()
    indices_to_change = np.random.choice(len(solution), size=int(0.1 * len(solution)), replace=False)
    perturbed_solution[indices_to_change] = 1 - perturbed_solution[indices_to_change]
    return perturbed_solution

# Função para o Simulated Annealing
def simulated_annealing(initial_solution, cost_function, temperature, cooling_rate, iterations):
    current_solution = initial_solution
    current_cost = cost_function(current_solution)

    best_solution = current_solution
    best_cost = current_cost

    temperature_list = []

    for i in range(iterations):
        temperature_list.append(temperature)

        next_solution = perturb_solution(current_solution)
        next_cost = cost_function(next_solution)

        if next_cost > current_cost or np.random.rand() < np.exp((current_cost - next_cost) / temperature):
            current_solution = next_solution
            current_cost = next_cost

            if current_cost > best_cost:
                best_solution = current_solution
                best_cost = current_cost

        temperature *= cooling_rate

    return best_solution, best_cost, temperature_list

# Parâmetros do Simulated Annealing
np.random.seed(42)  # Define a seed do gerador de números aleatórios
initial_temperature = 1000.0
cooling_rate = 0.99
iterations = 1000

# Executa o Simulated Annealing
utility, capacity = create_ksp(100)
choice = np.random.choice([0, 1], 100)
max_capacity = 2500

# Função de custo sem negação
def cost_function(solution):
    return objective_function(utility, capacity, max_capacity, solution)

best_solution, best_cost, temperature_list = simulated_annealing(
    choice, cost_function,
    initial_temperature, cooling_rate, iterations
)


# Testa várias combinações de Temperatura Inicial e cooling rate
initial_temperatures = [100, 1000, 5000]
cooling_rates = [0.95, 0.99, 0.90]

for initial_temperature in initial_temperatures:
    for cooling_rate in cooling_rates:
        # Executa o Simulated Annealing
        best_solution, best_cost, temperature_list = simulated_annealing(
            choice, cost_function,
            initial_temperature, cooling_rate, iterations
        )

        print(f"Resultados para Temperatura Inicial {initial_temperature} e Cooling Rate {cooling_rate}:")
        print("Melhor solução encontrada:", best_solution)
        print("Valor da melhor solução:", best_cost)
        total_weight_of_best_solution = np.dot(capacity, best_solution)
        print("Peso total da melhor solução:", total_weight_of_best_solution)
        print("\n")

C:\Users\Carlos\AppData\Local\Temp\ipykernel_24556\3452102772.py:43: RuntimeWarning: overflow encountered in exp
  if next_cost > current_cost or np.random.rand() < np.exp((current_cost - next_cost) / temperature):


Resultados para Temperatura Inicial 100 e Cooling Rate 0.95:
Melhor solução encontrada: [1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0 0 0 0 1 1 0 0 0 1 1 1 1 1 1 0 0 0 1 0
 0 0 0 0 0 0 1 1 1 0 1 1 1 1 1 0 0 1 0 1 0 0 0 0 1 0 0 1 0 1 0 1 1 0 1 0 0
 1 0 1 1 1 0 1 1 0 0 0 1 1 1 0 1 1 1 1 0 0 1 0 1 0 0]
Valor da melhor solução: 3044
Peso total da melhor solução: 2480


Resultados para Temperatura Inicial 100 e Cooling Rate 0.99:
Melhor solução encontrada: [0 1 0 1 1 1 1 1 0 1 1 0 1 1 0 1 0 0 0 0 1 0 0 1 1 1 1 0 1 1 1 1 1 1 1 1 0
 0 0 0 1 0 0 0 1 1 1 0 1 0 0 1 0 1 0 1 0 0 0 1 0 0 1 0 1 1 1 0 1 0 1 0 0 0
 1 0 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 0 1 1 1 1 1 1]
Valor da melhor solução: 3253
Peso total da melhor solução: 2392


Resultados para Temperatura Inicial 100 e Cooling Rate 0.9:
Melhor solução encontrada: [1 0 1 1 0 1 1 1 0 1 1 1 1 1 0 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 0 0 0 0 0 1
 0 1 0 1 0 1 0 1 1 1 0 1 1 1 0 1 0 0 0 0 1 1 0 1 1 1 0 0 1 1 0 1 1 1 0 0 1
 0 0 0 1 1 1 0 1 0 1 1 1 0 1 0 0 1 0 1 1 1 1 

Ao analisar os resultados obtidos para diferentes combinações de Temperatura Inicial e Cooling Rate no algoritmo Simulated Annealing, podemos tirar as seguintes conclusões:

-Impacto da Temperatura Inicial:

-Menor Temperatura Inicial (100): Pode levar a soluções iniciais menos exploradas. No entanto, pode-se ficar preso em ótimos locais, resultando em soluções subótimas.
-Maior Temperatura Inicial (1000, 5000): Proporciona uma exploração mais ampla do espaço de solução. Pode ajudar a escapar de ótimos locais, mas também pode ser mais propenso a aceitar soluções subótimas.


-Impacto do Cooling Rate:

-Cooling Rate mais baixo (0.90): Mantém a temperatura por mais tempo, permitindo uma exploração mais extensa do espaço de solução. Pode ser útil para evitar ficar preso em ótimos locais, mas pode levar a um aumento no tempo computacional.
-Cooling Rate mais Alto (0.99): Reduz a temperatura rapidamente, concentrando-se mais na exploração local. Pode ser eficaz para convergir rapidamente para uma solução, mas corre o risco de ficar preso em ótimos locais.


-Variação nos Resultados:

-É possível observar variação nos resultados entre diferentes execuções, mesmo com os mesmos parâmetros iniciais. Isso é comum em algoritmos estocásticos como o Simulated Annealing. 

-Desempenho das Soluções Encontradas:

-As soluções encontradas têm valores de utilidade variados, indicando diferentes níveis de eficiência em termos de utilidade total. O peso total da melhor solução também varia, indicando diferentes capacidades atendidas.
Em suma, é possível ajustar os parâmetros do Simulated Annealing com base em preferências pessoais e nos requisitos específicos do problema. Experimentar diferentes combinações de Temperatura Inicial e Cooling Rate pode ajudar a encontrar um equilíbrio entre a exploração e a exploração local, resultando em soluções de alta qualidade.

3. O que acontece se o item adicionado for aquele que tiver um rácio de utilidade e peso maior em todas as iterações? A solução final com este método é melhor que a solução original?

Aplicamos o algoritmo "greedy" para resolver esta questão:

In [ ]:
#################   3. O que acontece se o item adicionado for aquele que tiver um rácio de utilidade e peso   ##################
################# maior em todas as iterações? A solução final com este método é melhor que a solução original?  #################


import numpy as np

# Criação do dataset
def create_ksp(number_of_items, max_utility=100, max_capacity=100):
    utility = np.random.randint(0, max_utility, number_of_items)
    capacity = np.random.randint(1, max_capacity, number_of_items)  # Garante que o peso não seja zero
    return utility, capacity

# Função para resolver o problema da mochila com a estratégia de escolha do maior rácio de valor e peso (algoritmo greedy)
def solve_ksp_greedy_ratio(utility, capacity, max_capacity):
    num_items = len(utility)
    ratios = utility / capacity  # Calcula o rácio de utilidade e peso para cada item
    sorted_indices = np.argsort(ratios)[::-1]  # Ordena os índices pelo rácio em ordem decrescente

    current_capacity = 0
    solution = np.zeros(num_items)

    for index in sorted_indices:
        if current_capacity + capacity[index] <= max_capacity:
            solution[index] = 1
            current_capacity += capacity[index]

    return solution  # Devolve apenas a solução

# Criação do problema da mochila
np.random.seed(10)
utility, capacity = create_ksp(100)
max_capacity = 2500

# Resolve o problema da mochila usando a estratégia do maior rácio de utilidade e peso
greedy_solution = solve_ksp_greedy_ratio(utility, capacity, max_capacity)

# Calcula o peso total da solução
total_capacity = np.dot(capacity, greedy_solution)

# Calcula o valor total da solução
total_utility = np.dot(utility, greedy_solution)

# Mostra os resultados
print("Valor dos itens:", utility)
print("Peso dos itens:", capacity)
print("Capacidade máxima da mochila:", max_capacity)
print("Solução encontrada pelo algoritmo greedy:", greedy_solution)
print("Valor total da solução rácio:", total_utility)
print("Peso total da solução rácio:", total_capacity)

Utilidade dos itens: [ 9 15 64 28 89 93 29  8 73  0 40 36 16 11 54 88 62 33 72 78 49 51 54 77
 69 13 25 13 92 86 30 30 89 12 65 31 57 36 27 18 93 77 22 23 94 11 28 74
 88  9 15 18 80 71 88 11 17 46  7 75 28 33 84 96 88 44  5  4 71 88 88 50
 54 34 15 77 88 15  6 85 22 11 12 92 96 62 57 79 42 57 97 50 45 40 89 73
 37  0 18 23]
Peso dos itens: [ 4 30 17 85 83 15 52 80 18 51 54 26 49 18 33 82 81 42 91 13 31 82 18 17
  1 32 74 65 39 23 97 67 68 63 96 28 83 63 78 49 94 76 87 38 12 22 34 96
 44 89 97 74 41 44 91 72  9 86 73 29 31 90 26 79 82 86 63 14 42 34  5 88
 95 29 40 92 10  8 23 33  4 10 53 77 69 31 71 75 31 10  3 66 14 76 53  6
 94 85 49 63]
Capacidade máxima da mochila: 2500
Solução encontrada pelo algoritmo greedy: [1. 0. 1. 0. 1. 1. 0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1.
 1. 0. 0. 0. 1. 1. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 1. 1. 0. 1. 1. 0. 1. 1.
 1. 0. 0. 0. 1. 1. 1. 0. 1. 0. 0. 1. 1. 0. 1. 1. 1. 0. 0. 1. 1. 1. 1. 0.
 0. 1. 0. 1. 1. 1. 0. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 

Comparando os resultados entre a solução encontrada pelo algoritmo greedy, que utiliza a estratégia do maior rácio de utilidade e peso,
e a solução original do Simulated Annealing, podemos observar algumas conclusões:

Valor Total da Solução:
Greedy: 3932
Simulated Annealing: 3103
O algoritmo greedy, ao escolher itens com base no maior rácio de utilidade e peso, obteve um valor total de solução maior em comparação com o Simulated Annealing.

    
Peso Total da Solução:
Greedy: 2498
Simulated Annealing: 2477
O algoritmo greedy gerou uma solução com um peso total ligeiramente maior em comparação com o Simulated Annealing.

Conclusões:

O algoritmo greedy tende a favorecer itens com rácios mais altos de valor/peso, o que pode levar a soluções com valores totais mais altos.
O Simulated Annealing, pela sua natureza estocástica, pode explorar diferentes combinações de itens, resultando em soluções que podem não ser 
estritamente baseadas no rácio de valor e peso. Essas diferenças destacam a natureza das estratégias dos algoritmos. O algoritmo greedy foca-se
em escolhas locais ótimas em cada passo,enquanto o Simulated Annealing procura explorar diferentes regiões do espaço de solução, permitindo
a aceitação ocasional de soluções piores para evitar ficar preso em ótimos locais.
Em diferentes contextos ou requisitos de otimização, uma abordagem pode ser mais adequada do que a outra.